lib/SILOptimizer/Mandatory/Differentiation.cpp

```diff
@@ -2151,6 +2214,25 @@ void DifferentiableActivityInfo::propagateVariedInwardsThroughProjections(
   auto *inst = value->getDefiningInstruction();
   if (!inst || isa<ApplyInst>(inst))
     return;
+  // Recursively propagate varied through users that are projections or
```

Comment:
Shouldn't the normal `propagateVaried` flow handle everything that this new block of code does, because this is just propagation from operands to results?

Reply:
Oh, perhaps this is currently necessary because you're calling `setVaried()` a few lines above instead of `setVariedAndPropagateToUsers()`? If changing it to `setVariedAndPropagateToUsers()` is correct, then that seems like a cleaner simpler solution.

Reply:
Good suggestion! I made changes in https://github.com/apple/swift/pull/28155/commits/2a0e377133dff15209fe9dbc4923366051a52bc6:
* `propagateVariedInwardsThroughProjections` now calls `setVariedAndPropagateToUsers`, simplifying code.
* Special address-projection logic was removed. It seems no longer necessary.

It's still necessary to keep `propagateVariedInwardsThroughProjections` because it needs to recurse on instruction operands.